src/core/src/op/loop.cpp

```diff
@@ -191,6 +191,66 @@ void op::v5::Loop::validate_and_infer_types() {
     // Body
     m_bodies[0]->validate_nodes_and_infer_types();
 
+    // if output shape is not same with input in a back-edge, here will update the input shape
+    // Port map processing: output -> input
+    std::map<uint64_t, uint64_t> back_edges;
+    for (const auto& desc : get_input_descriptions()) {
+        if (const auto& merged_desc = std::dynamic_pointer_cast<opset5::Loop::MergedInputDescription>(desc)) {
+            back_edges[merged_desc->m_body_value_index] = merged_desc->m_body_parameter_index;
+        }
+    }
+    if (!back_edges.empty()) {
+        bool need_reinvalidate = false;
+        for (const auto& output_description : m_output_descriptions[0]) {
+            auto body_value = m_bodies[0]->get_results().at(output_description->m_body_value_index)->input_value(0);
+
+            if (auto body_output_description =
+                    ov::as_type_ptr<v0::TensorIterator::BodyOutputDescription>(output_description)) {
+                const ov::PartialShape& body_value_shape = body_value.get_partial_shape();
+                if (body_value_shape.rank().is_static()) {
+                    // handle the case: when sub-model's output shape does not compatible input shape in a back-edge,
+                    // such as
+                    //          Parameter(out:-1, 1)->|
+                    //                                |->Concat(out:-1, 2)->Result(out:-1, 2)
+                    //          Parameter(out:-1, 1)->|   (axis==1)
+                    // when iteration number is unknown or sub-model output shape may be vary, the Result shape should
+                    // infer as (-1, -1), then set changed one to input and propagate to others.
+                    if (back_edges.count(output_description->m_body_value_index)) {
+                        auto input_param =
+                            m_bodies[0]->get_parameters().at(back_edges[output_description->m_body_value_index]);
+                        const auto& input_param_ps = input_param->get_partial_shape();
+                        if (input_param_ps.rank().is_static()) {
+                            const auto body_rank_len = body_value_shape.rank().get_length();
+                            const auto input_rank_len = input_param_ps.rank().get_length();
+                            // if input and output shape are {n} or {} and not compatible, make output shape dynamic
+                            if (body_rank_len <= 1 && input_rank_len <= 1 &&
+                                !input_param_ps.compatible(body_value_shape)) {
+                                // reset sub model input shape
+                                input_param->set_partial_shape({-1});
```

Comment:
If both input and output shape are `[]` (scalars), we are setting new shape with rank 1: `[-1]`. So we are changing rank from 0 to 1, it doesn't look correct.

Reply:
I've checked the Concat [spec](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/movement/Concat_1.md) which does not allow scalar inputs. Currently the following graph should not be possible:
<pre>
Parameter(scalar)            Const(scalar)
       |_________________________|
                |
            Concat
                |
           Result(back edge to the Parameter)
</pre>
And I think there should also not exist the similar subgraph. @mbencer If you have any concern please comment, thanks.
I will remove the logic.

Reply:
why do we need to handle this case separately? why can't we use  `body_rank_len == input_rank_len && !input_param_ps.compatible(body_value_shape))` branch ?

Reply:
The scalar path removed, all use `body_rank_len == input_rank_len && !input_param_ps.compatible(body_value_shape))` branch.